presto-orc/src/main/java/com/facebook/presto/orc/metadata/ColumnEncoding.java

```diff
@@ -25,15 +27,25 @@
         DIRECT_V2,
         DICTIONARY_V2,
         DWRF_DIRECT,
+        FLATTENED,
```

Comment:
Thanks for taking a look Wenlei, regarding the Optional<List<DwrfFlattenMapEncoding>> within ColumnEncoding, I had originally started with an approach like that, but then I discovered a bunch of places in the code where we iterate over the ColumnEncodings to do something, e.g. in StripeReader to check if any columns are dictionary encoded, as well as when creating value streams, reading the indexes, creating checkpoints, etc.

I was worried that if I effectively hid these additional ColumnEncodings, it would be easy to leave that out in any new code that did something similar.  Especially given the niche nature of the flattened map feature, many contributors wouldn't be aware of it and it would be easy to miss in testing especially for new code not covered by existing tests.

That's why I opted to instead make it very "in your face" so that it's difficult (though unfortunately still not impossible) to accidentally miss.

Ideally there would be some abstraction that makes it so contributors won't need to be aware of it at all unless they're working on flattened maps specifically.  I wasn't unable to come up with one, but I'd love to discuss it.

Reply:
will do